utils_cv/detection/dataset.py

```diff
@@ -0,0 +1,220 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import os
+import math
+from pathlib import Path
+from random import randrange
+from typing import List, Tuple, Union
+
+import torch
+from torch.utils.data import Dataset, Subset
+import xml.etree.ElementTree as ET
+from PIL import Image
+
+from .plot import display_bounding_boxes, plot_grid
+from .model import get_transform
+
+
+class DetectionDataset(object):
+    """ An object detection dataset.
+    """
+
+    def __init__(
+        self,
+        root: Union[str, Path],
+        transforms: object = None,
+        image_folder: str = "images",
+        annotation_folder: str = "annotations",
+    ):
+        """ initialize dataset
+
+        This class assumes that the data is formatted in two folders:
+            - annotation folder which contains the Pascal VOC formatted
+              annotations
+            - image folder which contains the images
+
+        Args:
+            root: the root path of the dataset containing the image and
+            annotation folders
+            transforms: the transformations to apply
+            image_folder: the name of the image folder
+            annotation_folder: the name of the annotation folder
+        """
+
+        self.root = Path(root)
+        self.transforms = transforms
+        self.image_folder = image_folder
+        self.annotation_folder = annotation_folder
+
+        self.ims = list(sorted(os.listdir(self.root / self.image_folder)))
+        self.annotations = list(
+            sorted(os.listdir(self.root / self.annotation_folder))
+        )
+        self.categories = self._get_categories()
+
+    def _get_categories(self) -> List[str]:
+        """ Parses all Pascal VOC formatted annoatation files to extract all
+        possible categories. """
+        categories = ["__background__"]
+        for annotation_path in self.annotations:
+            annotation_path = self.root / "annotations" / str(annotation_path)
+            tree = ET.parse(annotation_path)
+            root = tree.getroot()
+            objs = root.findall("object")
+            for obj in objs:
+                category = obj[0]
+                assert category.tag == "name"
+                categories.append(category.text)
+        return list(set(categories))
+
+    def get_image_features(
```

Comment:
"features" is unclear and a bit misleading. Maybe name get_annos() or get_anno_bboxes()   (see comment below what I mean with the BBox class)